File: cpp-answers/112.路径总和.cpp
/*
 * @Author: your name
 * @Date: 2020-06-03 13:42:41
 * @LastEditTime: 2020-07-07 17:14:46
 * @LastEditors: Hongyang_Yang
 * @Description: In User Settings Edit
 * @FilePath: \postwoman-mastere:\leetcode刷题\112.路径总和.cpp
 */
/*
 * @lc app=leetcode.cn id=112 lang=cpp
 *
 * [112] 路径总和
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
/*递归/深度优先
class Solution
{
public:
    bool hasPathSum(TreeNode *root, int sum)
    {
        if (root == NULL)
            return false;
        int newsum = sum - root->val;
        if (root->left == NULL && root->right == NULL)
        {
            return newsum == 0;
        }
        return hasPathSum(root->left, newsum) || hasPathSum(root->right, newsum);
    }
};*/
/*广度优先搜索*/
class Solution
{
public:
    bool hasPathSum(TreeNode *root, int sum)
    {
        if (root == NULL)
            return false;
        queue<TreeNode *> que_node;
        queue<int> que_val;
        que_node.push(root);
        que_val.push(root->val);
        while (!que_node.empty())
        {
            TreeNode *now = que_node.front();
            int temp = que_val.front();
            que_node.pop();
            que_val.pop();
            if (now->left == nullptr && now->right == nullptr)
            {
                if (temp == sum)
                    return true;
                continue;
            }
            if (now->left != nullptr)
            {
                que_node.push(now->left);
                que_val.push(now->left->val + temp);
            }
            if (now->right != nullptr)
            {
                que_node.push(now->right);
                que_val.push(now->right->val + temp);
            }
        }
        return false;
    }
};
// @lc code=end

File: rust-answers/112.路径总和.rs
/*
 * @lc app=leetcode.cn id=112 lang=rust
 *
 * [112] 路径总和
 */

// @lc code=start
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
    pub fn has_path_sum(root: Option<Rc<RefCell<TreeNode>>>, target_sum: i32) -> bool {
        if let Some(n) = root {
            let node = n.borrow();
            if node.left.is_none() && node.right.is_none() && node.val == target_sum {
                return true;
            } 
            return Solution::has_path_sum(node.left.clone(), target_sum - node.val) || Solution::has_path_sum(node.right.clone(), target_sum - node.val)
        }
        return false;
    }
}
// @lc code=end

